### DISCUSSION TITLE
Radiative Heat Flux as a Kernel

### DISCUSSION THREAD

**Discussion by Zoophish** [idaholab/moose#24467] [Q&A Modules: General] [answered]
Hi,

We have been trying to implement radiative cooling on the interface between two different materials, with different material properties, that are in the same block.
This is so that elements can change variables without having to move the element between blocks, which is the major contributor to the run time.
 For our application, we have chosen to distinguish between the two materials using an additional 'switch' variable which is either 1 or 0 on the nodes rather than using two blocks. 
We also change the material properties as a function of their position. 

Since we no longer have a sideset to apply a BC to, we wanted to see if it was possible to apply an equivalent radiative heat flux kernel that radiates heat from nodes that are approximately at the materials interface.

To do this, we have implemented the following kernel based on the RadiativeHeatFluxBase BC from the Heat Conduction module. This only returns a non zero residual contribution if the switch value is between the threshold values given; in this case those thresholds are 0.1 and 0.9.

```cpp
#pragma once

#include "RadiationCooling2.h"

#include "Function.h"

#include "MathUtils.h"

registerMooseObject("MyApp", RadiationCooling2);

InputParameters
RadiationCooling2::validParams()
{
   InputParameters params = ADKernelValue::validParams();
   params.addClassDescription(
       "Apply radiative cooling to the boundary of phi (phi = 0.5). This acts as a heat sink so takes energy out of the model.");
   params.addRequiredCoupledVar("switch", "switch value");
   params.addParam<Real>("thresh", 0.1, "Threshold for the gradient value");
   // params.addRequiredCoupledVar("phi", "The level set variable");
   params.addParam<Real>("ambient_temperature", 300, "The ambient temperature");
   // params.addRequiredParam<Real>("sigma", "The standard deviation of the Gaussian around the interface");
   params.addParam<Real>("factor", 1.0, "Scaling coefficient to multiply the heat sink by");
   params.addParam<MaterialPropertyName>("emissivity", "emissivity", "The name of the emissivity material.");
   return params;
}

RadiationCooling2::RadiationCooling2(const InputParameters & parameters)
 : ADKernelValue(parameters),
 _switch(adCoupledValue("switch")),
 _thresh(getParam<Real>("thresh")),
 _T0(getParam<Real>("ambient_temperature")),
 _emissivity(getADMaterialProperty<Real>("emissivity")),
 _factor(getParam<Real>("factor"))
{
}

ADReal
RadiationCooling2::precomputeQpResidual()
{
   if((_switch[_qp] > _thresh) && (_switch[_qp] < (1-_thresh))){  // new

       ADReal heatFlux = 5.670367e-14 * std::max(std::pow(_u[_qp], 4) - std::pow(_T0, 4), 0.0); // TempDiff always greater than 0
     return  _factor * _emissivity[_qp] * heatFlux;
   }
   else{

       return 0;
   }
}
```

We have tried this on a simple 2D test case consisting of a metal square (switch=1) within a larger air square (switch=0). We also created an equivalent metal square with a FunctionRadiativeBC on the external boundary and compared the temperatures of both the models over time.

From our tests we have found that:
The cooling needs to be multiplied by a 1/element length. This has been tested with different resolutions and has been shown to hold.
All of these results are still out by a factor of ~1.2, when testing with squares of different resolutions in 2D, and also cubes in 3D, as seen below.

![image](https://github.com/idaholab/moose/assets/3082590/a6cdd309-4839-4c75-a54d-248efe041608)

However, when trying to validate with a circle instead, the factor of 1.2 is not correct.

My questions:
Is this a feasible approach to take?
Do you have any ideas where the factors may have come from?
Is there a better way to do it?

Thanks for your time

**Comment by hugary1995** [comment]
I don't think the math works out like that. You are basically approximating a boundary integral (the original radiation BC) with a volume integral (using the switch function). Typically, people in the phase field community use an indicator function to do the regularization. See section 2.1 of "A phase-field model of quasi-brittle fracture for pressurized cracks: Application to UO
 high-burnup microstructure fragmentation". The derivation tells you how to do the approximation with gamma convergence.

**Reply by Zoophish** [reply]
Okay that would explain our results, I'll take a look at that paper. Thanks for your help

**Reply by lloyd-5731** [reply]
Hi @hugary1995 , I have had a look through the paper you recommended, thank you for that. I can't seem to understand how this would be implemented into MOOSE at the minute.

Would you be able to explain any further, as I think this paper is currently slightly beyond my understanding.
If I share my current kernel would you be able to explain the implementation please?

All the best,
Lloyd

**Reply by hugary1995** [reply, marked as answer]
I wrote down some equations to explain.
[scratch_paper.pdf](https://github.com/idaholab/moose/files/11614281/scratch_paper.pdf)

Eq. 4 is what you need to implement. OP was only missing that indicator function. Though you do need to figure out how to construct/solve/update the order parameter. If you have a level-set, that would simplify things.

**Reply by hugary1995** [reply]
I suspect MALAMUTE already has something like this. @jiangwen84 would know.

**Reply by lloyd-5731** [reply]
Thank you very much. 
I have found the relevant MALAMUTE code now also: https://github.com/idaholab/malamute/blob/devel/src/kernels/MeltPoolHeatSource.C

I think the only step I am missing now is now to implement the Indicator function/ level set.

In eq4 is it the indicator function or the magnitude of the gradient of the indicator function that you integrate over?

Thanks again,
Lloyd

**Reply by hugary1995** [reply]
oops, that was a typo. It should be magnitude of the gradient of the indicator function.

**Reply by Zoophish** [reply]
Thank you very much for your help, I have made a working solution now based on your explanation.

**Reply by hugary1995** [reply]
Fantastic!
